File: tracking-program/8087/api/services/utilities/equipment-calculations.js
module.exports = {
  calculate: function(project, metersToReport) {
    let equipments = {};
    console.log("start calculator");
  
    if (!metersToReport) {
        console.log("!metersToReport");
        console.log(project.equipmentInfo);
        if (project.equipmentInfo.meterEquipment == null) {
            equipments = project.equipmentInfo;
        } else {
            equipments = project.equipmentInfo.meterEquipment;
        }
    } else {
	    console.log("metersToReport");
        equipments = project.equipmentInfo.meterEquipment.filter(meter => { return metersToReport.includes(meter.meterNumber)});
    }

    console.log("st2");
    if (!equipments) { 
        console.log("return project.equipmentInfo");
        return project.equipmentInfo;
    }

    console.log("st3");
    let equipmentInfo = "";
    if (equipments.length > 1) { //first one is cascade wtf
        console.log("equipments[1]:",equipments[1]);
        equipmentInfo = equipments[1];
    } else {
        console.log("equipments:",equipments);
        equipmentInfo = equipments;
    }

    console.log("st4");
    console.log(equipmentInfo);

    //equipmentInfo.total.discount = equipmentInfo.total.itemTotal * project.discount / 100;
    console.log("st5");
    if (equipments.length > 2) {
    console.log("st6");
  
        for (i = 2; i < equipments.length; i++) {
            let services = [];
            console.log("equipments[",i,"].services:", equipments[i].services);
            equipments[i].services.forEach(function(service) {
                let tempService = equipmentInfo.services.find(serv => {return serv.name == service.name});
                tempService.price = parseFloat(tempService.price) + parseFloat(service.price);
                services.push(tempService);
            });

            equipmentInfo.services = services;
            console.log("equipmentInfo.services", equipmentInfo.services);
            let parts = [];
            console.log("equipments[",i,"].parts:", equipments[i].parts);
            equipments[i].parts.forEach(function(part) {
                let tempPart = equipmentInfo.parts.find(p => {return p.name == part.name});
                tempPart.count = parseInt(tempPart.count) + parseInt(part.count);
                //console.log("tempPart.count += part.count");
                //console.log(tempPart.count);
                parts.push(tempPart);
            });

            equipmentInfo.parts = parts;
            console.log("equipmentInfo.parts", equipmentInfo.parts);
            let items = [];
            console.log("equipments[",i,"].items:", equipments[i].items);
            equipments[i].items.forEach(function(item) {
                let tempItem = equipmentInfo.items.find(it => {return it.name == item.name});
                tempItem.count = parseInt(tempItem.count) + parseInt(item.count);
                items.push(tempItem);
            });
            equipmentInfo.items = items;
            console.log("equipmentInfo.items", equipmentInfo.items);

            equipmentInfo.total.itemTotal = equipments[i].total.itemTotal + equipmentInfo.total.itemTotal;
            equipmentInfo.total.subtotal = equipments[i].total.subtotal + equipmentInfo.total.subtotal;
            equipmentInfo.total.tax = equipments[i].total.tax + equipmentInfo.total.tax;
            //equipmentInfo.total.discount = equipments[i].total.discount + equipmentInfo.total.discount;
            equipmentInfo.total.discount = (equipmentInfo.total.itemTotal * project.discount / 100) + equipmentInfo.total.discount;
            equipmentInfo.total.total = equipments[i].total.total + equipmentInfo.total.total;
            console.log("equipmentInfo.total.total", equipmentInfo.total.total);
            //console.log(equipmentInfo);
        }
    }
    console.log("Equipments Calculator Done");
    return equipmentInfo;
  }
}

File: tracking-program/8087/flask_app/app/services/equipment_calculations.py
"""
Equipment calculations - port of api/services/utilities/equipment-calculations.js
Selects/aggregates equipmentInfo from meterEquipment when metersToReport filters meters.
"""
import copy


def calculate(project, meters_to_report=None):
    equip = (project.equipmentInfo or {}) if hasattr(project, "equipmentInfo") else {}
    meter_equip = equip.get("meterEquipment")

    if not meters_to_report:
        if meter_equip is None:
            return equip
        equipments = meter_equip
    else:
        if not meter_equip or not isinstance(meter_equip, list):
            return equip
        equipments = [m for m in meter_equip if m.get("meterNumber") in meters_to_report]
        if not equipments:
            return equip

    if not isinstance(equipments, list):
        return equipments
    if not equipments:
        return equip

    if len(equipments) > 1:
        equipment_info = copy.deepcopy(equipments[1])
    else:
        equipment_info = copy.deepcopy(equipments[0])

    if len(equipments) <= 2:
        return equipment_info

    equipment_info.setdefault("items", [])
    equipment_info.setdefault("parts", [])
    equipment_info.setdefault("services", [])
    equipment_info.setdefault("total", {"itemTotal": 0, "subtotal": 0, "tax": 0, "discount": 0, "total": 0})
    discount_pct = float(getattr(project, "discount", 0) or 0) / 100

    for i in range(2, len(equipments)):
        other = equipments[i]
        for svc in (other.get("services") or []):
            existing = next((s for s in equipment_info["services"] if s.get("name") == svc.get("name")), None)
            if existing:
                existing["price"] = float(existing.get("price") or 0) + float(svc.get("price") or 0)
        for p in (other.get("parts") or []):
            existing = next((x for x in equipment_info["parts"] if x.get("name") == p.get("name")), None)
            if existing:
                existing["count"] = int(existing.get("count") or 0) + int(p.get("count") or 0)
        for it in (other.get("items") or []):
            existing = next((x for x in equipment_info["items"] if x.get("name") == it.get("name")), None)
            if existing:
                existing["count"] = int(existing.get("count") or 0) + int(it.get("count") or 0)
        t = equipment_info["total"]
        ot = other.get("total") or {}
        t["itemTotal"] = float(t.get("itemTotal") or 0) + float(ot.get("itemTotal") or 0)
        t["subtotal"] = float(t.get("subtotal") or 0) + float(ot.get("subtotal") or 0)
        t["tax"] = float(t.get("tax") or 0) + float(ot.get("tax") or 0)
        t["discount"] = (t["itemTotal"] * discount_pct) + float(t.get("discount") or 0)
        t["total"] = float(t.get("total") or 0) + float(ot.get("total") or 0)

    return equipment_info
